prusti-viper/src/encoder/mir/specifications/interface.rs

```diff
@@ -0,0 +1,120 @@
+use crate::encoder::mir::specifications::specs::Specifications;
+use log::trace;
+use prusti_interface::specs::{typed, typed::DefSpecificationMap};
+use rustc_hash::FxHashMap;
+use rustc_hir::def_id::{DefId, LocalDefId};
+use std::cell::RefCell;
+
+pub(crate) struct SpecificationsState {
+    specs: RefCell<Specifications>,
+    pure_cache: RefCell<FxHashMap<DefId, bool>>,
+    trusted_cache: RefCell<FxHashMap<DefId, bool>>,
+}
+
+impl SpecificationsState {
+    pub fn new(user_typed_specs: DefSpecificationMap) -> Self {
+        Self {
+            specs: RefCell::new(Specifications::new(user_typed_specs)),
+            pure_cache: RefCell::default(),
+            trusted_cache: RefCell::default(),
+        }
+    }
+}
+
+pub(crate) trait SpecificationsInterface {
+    fn is_pure(&self, def_id: DefId) -> bool;
+
+    fn is_trusted(&self, def_id: DefId) -> bool;
+
+    fn get_predicate_body(&self, def_id: DefId) -> Option<LocalDefId>;
+
+    fn has_extern_spec(&self, def_id: DefId) -> bool;
+
+    /// Get the loop invariant attached to a function with a
+    /// `prusti::loop_body_invariant_spec` attribute.
+    fn get_loop_specs(&self, def_id: DefId) -> Option<typed::LoopSpecification>;
+
+    /// Get the specifications attached to the `def_id` function.
+    fn get_procedure_specs(&self, def_id: DefId) -> Option<typed::ProcedureSpecification>;
+
+    /// Get a local wrapper `DefId` for functions that have external specs.
+    /// Return the original `DefId` for everything else.
+    fn get_wrapper_def_id(&self, def_id: DefId) -> DefId;
+}
+
+impl<'v, 'tcx: 'v> SpecificationsInterface for super::super::super::Encoder<'v, 'tcx> {
+    fn is_pure(&self, def_id: DefId) -> bool {
+        let mut cache = self.specifications_state.pure_cache.borrow_mut();
+        let result = cache.entry(def_id).or_insert_with(|| {
+            self.specifications_state
+                .specs
+                .borrow_mut()
+                .compute_is_pure(self.env(), def_id)
+        });
+
+        trace!("is_pure {:?} = {}", def_id, result);
+        *result
+    }
+
+    fn is_trusted(&self, def_id: DefId) -> bool {
+        let mut cache = self.specifications_state.trusted_cache.borrow_mut();
+        let result = cache.entry(def_id).or_insert_with(|| {
+            self.specifications_state
+                .specs
+                .borrow_mut()
+                .compute_is_trusted(self.env(), def_id)
+        });
+        trace!("is_trusted {:?} = {}", def_id, result);
+        *result
+    }
+
+    fn get_predicate_body(&self, def_id: DefId) -> Option<LocalDefId> {
+        let mut specs = self.specifications_state.specs.borrow_mut();
```

Comment:
Does this need to be a local variable? Why not `.borrow_mut` in a chain like the other methods?

Reply:
Rust complains: `temporary value dropped while borrowed E0716`